### PR TITLE
Od/document c api

### PR DESCRIPTION
**Pull request by jajhall**
Added some more `enums` as `const HighsInt` in `highs_c_api.h`, and using them in `TestCAPI.c` and `call_highs_from_c.c`. Worked through the documentation in `highs_c_api.h`.

### CODE REVIEW

File: examples/call_highs_from_c.c [hunk 1]
```diff
@@ -62,14 +62,14 @@ void minimal_api() {
   // bound on x_1, it serves to illustrate a non-trivial packed
   // column-wise matrix.
   //
-  const int sense_minimization = 1;
-  const int sense_maximization = -1;
+  const int kHighsObjSenseMinimize = 1;
+  const int kHighsObjSenseMaximize = -1;
```
Comment (odow) [inline examples/call_highs_from_c.c]: These should be removed

File: src/interfaces/highs_c_api.h [hunk 2]
```diff
@@ -14,9 +14,74 @@
 #define HIGHS_C_API
 
 #include "util/HighsInt.h"
-const HighsInt HighsStatuskError = -1;
-const HighsInt HighsStatuskOk = 0;
-const HighsInt HighsStatuskWarning = 1;
+
+const HighsInt kHighsStatusError = -1;
```
Comment (odow) [inline src/interfaces/highs_c_api.h:18]: Two comments:
 * This is technically breaking for anyone who was using this in the C API
 * Why the `k` prefix? Is that standard somewhere? 

I had it as `[enum class][enum name]` which seemed to make sense to me. The class is `HighsStatus` and the enum is `kError`. It also makes all of the symbols exported by HiGHS in the C interface start with `Highs` which makes discoverability easier. But I'm not super attached to it, so `kHighs` is okay if you prefer.

File: src/interfaces/highs_c_api.h [hunk 3]
```diff
@@ -25,110 +90,115 @@ extern "C" {
...
- * @param aindex    array of length [numnz] with indices of matrix entries
- * @param avalue    array of length [numnz] with values of matrix entries
- *
- * @param colvalue       array of length [numcol], filled with the primal column
- *                       solution
- * @param coldual        array of length [numcol], filled with the dual column
+ *                  three arrays, `a_start`, `a_index`, and `a_value`. `a_start`
+ *                  is an array of length [num_col] containing the starting
+ * index of each column in `a_index`. If `a_format` is `kRowwise` the array is
+ * of length [num_row] corresponding to each row.
```
Comment (odow) [inline src/interfaces/highs_c_api.h]: Indentation is wrong

File: src/interfaces/highs_c_api.h [hunk 4]
```diff
@@ -25,110 +90,115 @@ extern "C" {
...
- *                       of the rows
- * @param modelstatus    termination status of the model after the solve
- *
- * @returns A non-zero return value indicates that a problem occured
- */
-HighsInt Highs_lpCall(const HighsInt numcol, const HighsInt numrow,
-                      const HighsInt numnz, const HighsInt a_format,
+ * @param col_basis_status  array of length [num_col], filled with the basis
+ *                          status of the columns in the form of a
+ * `HighsBasisStatus` enum
```
Comment (odow) [inline src/interfaces/highs_c_api.h]: Indentation is wrong

File: src/interfaces/highs_c_api.h [hunk 5]
```diff
@@ -25,110 +90,115 @@ extern "C" {
...
+ * @param row_basis_status  array of length [num_row], filled with the basis
+ *                          status of the rows in the form of a
+ * `HighsBasisStatus` enum
```
Comment (odow) [inline src/interfaces/highs_c_api.h]: Indentation is wrong

File: src/interfaces/highs_c_api.h [hunk 6]
```diff
@@ -25,110 +90,115 @@ extern "C" {
...
+ * @param model_status      termination status of the model after the solve in
+ * the form of a `HighsModel_Status` enum
```
Comment (odow) [inline src/interfaces/highs_c_api.h]: Indentation is wrong

File: src/interfaces/highs_c_api.h [hunk 7]
```diff
@@ -236,83 +307,90 @@ HighsInt Highs_passLp(void* highs, const HighsInt numcol, const HighsInt numrow,
...
+ * @param num_col     the number of columns
+ * @param num_row     the number of rows
+ * @param num_nz      the number of elements in the constraint matrix
+ * @param q_num_nz    the number of elements in the Hessian matrix
+ * @param a_format    the format of the constraint matrix to use in th form of a
+ *                    `HighsMatrixFormat` enum
+ * @param q_format    the format of the Hessian matrix to use in the form of a
+ *                    `HighsHessianFormat` enum
+ * @param sense       the optimization sense in the form of a `HighsObjSense`
+ * enum
```
Comment (odow) [inline src/interfaces/highs_c_api.h]: Indentation is wrong

File: src/interfaces/highs_c_api.h [hunk 8]
```diff
@@ -236,83 +307,90 @@ HighsInt Highs_passLp(void* highs, const HighsInt numcol, const HighsInt numrow,
...
+ * @param row_lower   array of length [num_row] with the upper row bounds
+ * @param row_upper   array of length [num_row] with the upper row bounds
+ * @param a_start     the constraint matrix is provided to HiGHS in compressed
+ *                    sparse column form (if `a_format` is `kColwise`, otherwise
+ *                    compressed sparse row form). The sparse matrix consists of
+ *                    three arrays, `a_start`, `a_index`, and `a_value`.
+ *                    `a_start` is an array of length [num_col] containing the
+ *                    starting index of each column in `a_index`. If `a_format`
+ *                    is `kRowwise` the array is of length [num_row]
+ * corresponding to each row.
```
Comment (odow) [inline src/interfaces/highs_c_api.h]: Indentation is wrong

File: src/interfaces/highs_c_api.h [hunk 9]
```diff
@@ -236,83 +307,90 @@ HighsInt Highs_passLp(void* highs, const HighsInt numcol, const HighsInt numrow,
...
+ * @param a_index     array of length [num_nz] with indices of matrix entries
+ * @param a_value     array of length [num_nz] with values of matrix entries
+ * @param q_start     the Hessian matrix is provided in the same format as the
+ *                    constraint matrix, using `q_start`, `q_index`, and
+ *                    `q_value` in the place of `a_start`, `a_index`, and
+ * `a_value`. If the model is linear, pass NULL.
```
Comment (odow) [inline src/interfaces/highs_c_api.h]: Indentation is wrong

File: src/interfaces/highs_c_api.h [hunk 10]
```diff
@@ -799,106 +878,107 @@ HighsInt Highs_addRow(void* highs, const double lower, const double upper,
...
 
 /**
  * Change the objective sense of the model.
  *
- * @param highs a pointer to the HiGHS model object
- * @param sense the new optimization sense. Use -1 for maximization and 1 for
- *              minimization.
+ * @param highs     a pointer to the Highs instance
+ * @param sense     the new optimization sense in the form of a `HighsObjSense`
+ * enum
```
Comment (odow) [inline src/interfaces/highs_c_api.h]: Indentation is wrong

File: src/interfaces/highs_c_api.h [hunk 11]
```diff
@@ -1441,85 +1524,62 @@ HighsInt Highs_getHessianNumNz(const void* highs);
...
- * @param col_value array of length [m] with optimal primal solution for each
- *                  column
- * @param col_dual  array of length [n] with optimal dual solution for each
- *                  column
- * @param row_dual  array of length [m] with optimal dual solution for each row
+ * @param highs      a pointer to the Highs instance
+ * @param num_col    the number of variables
+ * @param num_row    the number of rows
+ * @param col_value  array of length [num_col] with optimal primal solution for
+ * each column
```
Comment (odow) [inline src/interfaces/highs_c_api.h]: Indentation is wrong

File: src/interfaces/highs_c_api.h [hunk 12]
```diff
@@ -1441,85 +1524,62 @@ HighsInt Highs_getHessianNumNz(const void* highs);
...
+ * @param col_dual   array of length [num_col] with optimal dual solution for
+ * each column
```
Comment (odow) [inline src/interfaces/highs_c_api.h]: Indentation is wrong

File: src/interfaces/highs_c_api.h [hunk 13]
```diff
@@ -1441,85 +1524,62 @@ HighsInt Highs_getHessianNumNz(const void* highs);
...
+ * @param row_dual   array of length [num_row] with optimal dual solution for
+ * each row
```
Comment (odow) [inline src/interfaces/highs_c_api.h]: Indentation is wrong

File: src/interfaces/highs_c_api.h [hunk 14]
```diff
@@ -25,110 +90,115 @@ extern "C" {
...
  *
- * @param numcol    the number of columns
- * @param numrow    the number of rows
- * @param numnz     the number of nonzeros in the constraint matrix
- * @param a_format  the format of the constraint matrix
- * @param sense     the optimization sense. Use -1 for maximization, 1 otherwise
+ * @param num_col   the number of columns
+ * @param num_row   the number of rows
+ * @param num_nz    the number of nonzeros in the constraint matrix
+ * @param a_format  the format of the constraint matrix as a `HighsMatrixFormat`
```
Comment (odow) [inline src/interfaces/highs_c_api.h]:
```suggestion
 * @param a_format  the format of the constraint matrix as a `kHighsMatrixFormat`
```

We should use however we chose to name this. There are lots of other instances of enums as well.

**Review by odow — COMMENTED** [review]
Name changes look good. 

* I think I prefer `HighsStatuskError` as it translates more directly to the C++ API, but if we change to `kHighsStatusError` then there are some updates to be made through the docstrings.
* I tried to flag the indentation issues, but I might have missed some

**Comment by jajhall** [comment]
> Name changes look good.
> 
>     * I think I prefer `HighsStatuskError` as it translates more directly to the C++ API, but if we change to `kHighsStatusError` then there are some updates to be made through the docstrings.
> 
>     * I tried to flag the indentation issues, but I might have missed some
> Two comments:
>
>    * Why the k prefix? Is that standard somewhere?
>
> I had it as [enum class][enum name] which seemed to make sense to me. The class is HighsStatus and the enum is kError. It also makes all of the symbols exported by HiGHS in the C interface start with Highs which makes discoverability easier. But I'm not super attached to it, so kHighs is okay if you prefer.

I can see your logic, but the "k" still corrupts names, and HiGHS uses

https://google.github.io/styleguide/cppguide.html#Constant_Names

The "enums" in the C++ header files are named correctly in the docstrings. In C they aren't enums. Why don't the docstrings refer to the `const HighsInt` "families?" - or are these `const HighsInt` values considered to be enums in C?


>    * This is technically breaking for anyone who was using this in the C API

Indeed, and I see that they were in v1.1.1. We can keep these three, but document why they are different and indicate that they are deprecated

Sorry for all the indentation errors. Everything was in order until I ran `clang` formatting, which must have objected to some long lines. I'll fix these

**Comment by odow** [comment]
Ah I didn't know that about the Google style guide. Happy to go with that then.

> or are these const HighsInt values considered to be enums in C?

Perhaps `enum` was the wrong word in C. 

> I'll fix these tomorrow, since it's 22:00 here, now.

I'll take a pass at sorting the issues. I should have permission to push to this branch.

**Comment by jajhall** [comment]
Yes, @galabovaa championed the Google style guide. 
I've made a few changes just now, so I'll push them and leave you to it

**Comment by jajhall** [comment]
So, Windows doesn't like 

const HighsInt foo = 0;
const HighsInt bar = foo;

Live and learn!

**Review by odow — APPROVED** [review]
I think this is in a good state.

**Comment by jajhall** [comment]
Thanks. I'll merge once the tests go through. I've run `clang` formatting on my machine and it changes nothing

**Pull request merged**